test/circuit.js

```diff
@@ -0,0 +1,290 @@
+/* eslint max-nested-callbacks: ["error", 8] */
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+chai.use(dirtyChai)
+
+const parallel = require('async/parallel')
+const series = require('async/series')
+
+const isNode = require('detect-node')
+
+const utils = require('./utils/circuit')
+
+const proc = utils.setUpProcNode
+const js = utils.setUpJsNode
+const go = utils.setUpGoNode
+
+const ws = utils.wsAddr
+const star = utils.wsStarAddr
+const circuit = utils.circuitAddr
+
+const send = utils.send
+
+const base = '/ip4/127.0.0.1/tcp/0'
+
+const connect = (nodeA, nodeB, relay, timeout, callback) => {
+  if (typeof timeout === 'function') {
+    callback = timeout
+    timeout = 1000
+  }
+
+  series([
+    (cb) => nodeA.ipfsd.api.swarm.connect(ws(relay.addrs), cb),
+    (cb) => setTimeout(cb, timeout),
```

Comment:
Why Timeout? Swarm should just return after the connect is done.

Reply:
I've tried without the timeouts but it fails in a lot of cases. I believe it is related to WS somehow - but just a hunch.

Reply:
Thought about it some more and remembered the main reason this is there. Upon establishing a  muxed connection, circuit will try to send the `CAN_HOP` to the incoming peer to identify it as a relay, that _I believe_, takes time. I'm trying to think of a better way of confirming this, but it sounds plausible so far.

Reply:
That doesn't make sense. If there is any action caused and needed by swarm.connect, it won't finish (aka call the callback) until it's finished. Adding a timeout after the fact means that that swarm.connect is returning before the connection was finished, which is not the expected behavior.

Reply:
No, the action doesn't happen as part of the swarm connect. It uses `peer-mux-established` event to send the `CAN_HOP`, thats why the timeout works.

Reply:
It happens asynchronously, after swarm connect.

Reply:
@dryajov I see. What you mean is that the _receiving node_ will only know who is dialing after Identify finishes. That again is strange because the _starting node_ has to finish SECIO (which tells the other peer the PeerId) + Multiplex handshakes before swarm.connect returns.

Wanna draw how you think the dialing is working?

Reply:
I'll draw it out. But to clarify, all the `CAN_HOP` stuff happens after secio and multiplex has finished, it literally hooks into the `peer-mux-established` event. So this timeout is not for a particular `swarm.connect`, I just happened to put it in between the two connect calls, but in reality (and I've already changed that), it can pretty well go at the end of the two connects like so:

```
  series([
    (cb) => nodeA.ipfsd.api.swarm.connect(getWsAddr(relay.addrs), cb),
    (cb) => nodeB.ipfsd.api.swarm.connect(getWsAddr(relay.addrs), cb),
    (cb) => setTimeout(cb, timeout),
    (cb) => nodeA.ipfsd.api.swarm.connect(getCircuitAddr(nodeB.addrs), cb)
  ], callback)
```

It's there to give time for `nodeA` and `nodeB` to realize that `relay` is actually a `Relay` by sending it the `CAN_HOP` message, so that when `nodeA` dials to `nodeB` using the `/p2p-circuit` address, it knows that there is a relay they can use.

Reply:
I believe I understand where the confusion comes from. Those timeouts look like they are there for the `swarm.connect`, they are not - their purpose is to give time to circuit to send the `CAN_HOP` message, after the `swarm.connect` has finished. (Drawing on the way)

Reply:
So the flow is like this 

**First**:

`dial -> [low level transport] -> secio -> multiplex -> 'peer-mux-extablished' --> Muxed Connection` 

**Second**

Independent flow that happens inside circuit upon receiving the `peer-mux-extablished`

`on('peer-mux-extablished') -> send 'CAN_HOP' msg --> peer is/isn't a Relay`

Reply:
@dryajov shouldn't there be a call on libp2p that could tell you if a Peer that you are connected is indeed a good relay?

Reply:
Hmm, we could add the `CAN_HOP` into the dial flow if `circuit` is enabled - I just wonder what would be the benefit? Currently, relay discovery happens async, and allows `libp2p` to use a connection as soon as its valid, adding another step into that flow will slow that down by as long as it takes to identify the relay, and in most cases the relay might not be used. Maybe a better way is to emit a message that will announce that a relay has been discovered?

Reply:
BTW - adding that to the flow would not be an issue at all, we already have a method that does the discovery (`canHop`), it would just be a matter of calling it.

Reply:
Not that. I'm saying that PeerInfo -- https://github.com/libp2p/js-peer-info/blob/master/src/index.js -- should contain the information about the Protocol we know are supported and also if the node can be a relay, so that you can scan through your Peers set and see if the nodes you are connected are indeed relay nodes. This would remove any race condition.

Reply:
> should contain the information about the Protocol we know are supported 

The `PeerInfo` should already have all the multiaddrs that the peer is listening on, so that information is there. 

> and also if the node can be a relay, so that you can scan through your Peers set and see if the nodes you are connected are indeed relay nodes. 

AFAIU, this would imply having a flag in the `PeerInfo` announcing that the peer is a relay?Wouldn't this require changing go's `PeerInfo` as well?

Reply:
Yes. Better track this in a separate issue and add a TODO on the timeout that points to it :)

Reply:
Yeah, agree! We desperately need a sane service discovery mechanism...